mongo/integration/causal_consistency_test.go

```diff
@@ -0,0 +1,287 @@
+		_ = mongo.WithSession(mtest.Background, sess, func(sc mongo.SessionContext) error {
+			_ = mt.Coll.FindOne(sc, bson.D{})
+			return nil
+		})
+
+		level, sentOptime := getReadConcernFields(mt, mt.GetStartedEvent().Command)
+		assert.Equal(mt, "", level, "expected command to not have read concern level, got %s", level)
+		assert.NotNil(mt, sentOptime, "expected operation time on command, got nil")
+		assert.True(mt, currOptime.Equal(*sentOptime), "expected operation time %v, got %v", currOptime, sentOptime)
+	})
+	localRcOpts := options.Client().SetReadConcern(readconcern.Local())
+	mt.RunOpts("custom read concern", mtest.NewOptions().ClientOptions(localRcOpts), func(mt *mtest.T) {
+		sess, err := mt.Client.StartSession()
+		assert.Nil(mt, err, "StartSession error: %v", err)
+		defer sess.EndSession(mtest.Background)
+
+		_ = mongo.WithSession(mtest.Background, sess, func(sc mongo.SessionContext) error {
+			_ = mt.Coll.FindOne(sc, bson.D{})
+			return nil
+		})
+		currOptime := sess.OperationTime()
+		mt.ClearEvents()
+		_ = mongo.WithSession(mtest.Background, sess, func(sc mongo.SessionContext) error {
+			_ = mt.Coll.FindOne(sc, bson.D{})
+			return nil
+		})
+
+		level, sentOptime := getReadConcernFields(mt, mt.GetStartedEvent().Command)
+		assert.Equal(mt, "local", level, "expected read concern level 'local', got %s", level)
+		assert.NotNil(mt, sentOptime, "expected operation time on command, got nil")
+		assert.True(mt, currOptime.Equal(*sentOptime), "expected operation time %v, got %v", currOptime, sentOptime)
+	})
+	unackWcOpts := options.Collection().SetWriteConcern(writeconcern.New(writeconcern.W(0)))
+	mt.RunOpts("unacknowledged write", mtest.NewOptions().CollectionOptions(unackWcOpts), func(mt *mtest.T) {
+		// unacknowledged write should not update the operationTime of a session
+
+		sess, err := mt.Client.StartSession()
+		assert.Nil(mt, err, "StartSession error: %v", err)
+		defer sess.EndSession(mtest.Background)
+
+		_ = mongo.WithSession(mtest.Background, sess, func(sc mongo.SessionContext) error {
+			_, _ = mt.Coll.InsertOne(sc, bson.D{{"x", 1}})
+			return nil
+		})
+		optime := sess.OperationTime()
+		assert.Nil(mt, optime, "expected operation time nil, got %v", optime)
+	})
+	mt.Run("clusterTime included", func(mt *mtest.T) {
+		// $clusterTime should be included in commands if the deployment supports cluster times
+
+		_ = mt.Coll.FindOne(mtest.Background, bson.D{})
+		evt := mt.GetStartedEvent()
+		assert.Equal(mt, "find", evt.CommandName, "expected command 'find', got '%v'", evt.CommandName)
+		_, err := evt.Command.LookupErr("$clusterTime")
+		assert.Nil(mt, err, "expected $clusterTime in command, got nil")
+	})
+}
+
+func TestCausalConsistency_NotSupported(t *testing.T) {
+	// use RunOnBlock instead of mtest.NewOptions().MaxServerVersion("3.5").Topologies(mtest.Single) because
+	// these tests should be run on servers <= 3.5 OR standalones
+	rob := []mtest.RunOnBlock{
+		{MaxServerVersion: "3.5"},
```

Comment:
I don't get the max server version of 3.5.  3.5 is essentially the same as 3.6, so if a test shouldn't run on 3.6 it shouldn't run on 3.5 either.  Perhaps this should be 3.4 instead?

Reply:
Good catch. I thought putting `3.4` would actually skip 3.4 servers because the exact version would be something like `3.4.23` but turns out `compareVersions` already accounts for that. Changed in new patch.